crates/solver/src/settlement/settlement_encoder.rs

```diff
@@ -260,7 +334,15 @@ impl SettlementEncoder {
     pub fn finish(mut self) -> EncodedSettlement {
         self.drop_unnecessary_tokens_and_prices();
 
-        let clearing_prices = self
+        let (mut additional_tokens, mut additional_prices): (Vec<H160>, Vec<U256>) = self
+            .non_uniform_clearing_prices
+            .iter()
+            .map(|x| (x.token, x.price))
+            .unzip();
+        let uniform_clearing_price_vec_length = self.tokens.len();
+        let mut tokens = self.tokens.clone();
+        tokens.append(&mut additional_tokens);
+        let mut clearing_prices: Vec<U256> = self
```

Comment:
It is not obvious to me how the call in line 351 won't fail for tokens that are not in `self.clearing_prices` but are in `tokens` due to line 344. Could you please explain?

Reply:
Could it be that you're counting that every token traded by a liquidity order has an entry in the price vector? (I don't believe that must be the case)

Reply:
Good catch!
I moved the       `  tokens.append(&mut additional_tokens);` further down, then only the tokens for which we really have clearing prices are in the tokens vect.

Reply:
Isn't this deeper than that? I understand that you're counting on one of the prices to be in the clearing price vector, but that doesn't necessary needs to be the case, right?

Reply:
> I understand that you're counting on one of the prices to be in the clearing price vector, but that doesn't necessary needs to be the case, right?

I made an explicit comment about this. I think it's reasonable to have this assumption.
I would propose  to force the solvers giving the nominator price - the price in the sell token via the uniform cl earing price - as then the driver has one less degree of freedom: 
If the solvers would not give a price for the sell token, then the driver could choose [a*sell_price]/ [a*buy_price] with an arbitrary `a` as valid prices for the liquidity order. If we assume that the sell_price is always given, then the scaling factor is in the control of the solver.